### DISCUSSION TITLE
How to emit empty data

### DISCUSSION THREAD

**Discussion by BenLocal** [Totodore/socketioxide#213] [Q&A]
in js socketio can emit only event id data , like `.emit("init-room")`, then will send package like `42["init-room"]`. but in [socketioxide](https://github.com/Totodore/socketioxide) must send a `serde_json::Value` data. how can i emit be like js send a empty data

**Comment by Totodore** [comment]
Interesting, I didn't think about this.
As you can see in some examples you have to put your data in an array/tuple if you want to match the multi argument pattern in the js lib. More concretely something like this in js:
`socket.emit("event", "arg1", "arg2")` correspond to this with socketioxide  `socket.emit("event", ("arg1, "arg2"))`.

If you want to send nothing you can try to put an empty tuple `()` or an empty array `[]`.

**Reply by BenLocal** [reply]
I am try send tuple `()` or an empty array `[]`,but all not get expect package.
eg:
`emit("init-room", ())` => `["init-room", null]`
`emit("init-room", [(); 0])` => `["init-room", []]`

If `bin(bytes).emit("init-room", ())` or `bin(bytes).emit("init-room", [(); 0])` are sent together, the order in which the client obtains parameters may change.
eg in js socketio server: `.emit("init-room", bytes)`, js can send package `["init-room",{"_placeholder":true,"num":0}]`
and client can get args like: `on("init-room", (bytes) => { ... })`

**Reply by BenLocal** [reply]
I think the data type in `packet::PacketData::Event` and `packet::BinaryPacket` can change from `Value ` to `Option<Value>`
If Option::None is emit, and skip json serialize it

**Reply by tausifcreates** [reply]
This sounds like a Good idea to me. Let's see what Theo thinks about this.

**Comment by Totodore** [comment]
> If `bin(bytes).emit("init-room", ())` or `bin(bytes).emit("init-room", [(); 0])` are sent together, the order in which the client obtains parameters may change. eg in js socketio server: `.emit("init-room", bytes)`, js can send package `["init-room",{"_placeholder":true,"num":0}]` and client can get args like: `on("init-room", (bytes) => { ... })`

It is not very clear, it may help you to know that binary packets are always pushed at the _end_ of the payload. If someone would like to add this info to the doc feel free  to make a PR :).

> I am try send tuple () or an empty array [],but all not get expect package.
eg:
emit("init-room", [(); 0]) => ["init-room", []]

It is not possible to change from `Value` to `Option<Value>` to because the `emit` fn take a Serializable arg so it would force anyone to do `Some(data)` for each call. 
The only solution would be to switch from this `emit("init-room", [(); 0]) => ["init-room", []]` to this `emit("init-room", [(); 0]) => ["init-room"]` which is completely doable but then it would become impossible to put empty arrays. A better thing to do would be to replace this one `emit("init-room", ()) => ["init-room", null]` to `emit("init-room", ()) => ["init-room"]`.

In any case someone is going to complain that null / undefined / [] is not supported.

**Reply by tausifcreates** [reply]
The last line had me rolling 🤣
